llama_index/indices/multi_modal/base.py

```diff
@@ -0,0 +1,329 @@
+        nodes: Sequence[BaseNode],
+        show_progress: bool = False,
+        **insert_kwargs: Any,
+    ) -> None:
+        """Add document to index."""
+        if not nodes:
+            return
+
+        image_nodes: List[ImageNode] = []
+        text_nodes: List[BaseNode] = []
+
+        for node in nodes:
+            if isinstance(node, ImageNode):
+                image_nodes.append(node)
+            if node.text:
+                text_nodes.append(node)
+
+        # embed all nodes as text - incclude image nodes that have text attached
+        text_nodes = self._get_node_with_embedding(
+            text_nodes, show_progress, is_image=False
+        )
+        new_text_ids = self.storage_context.vector_stores[DEFAULT_VECTOR_STORE].add(
+            text_nodes, **insert_kwargs
+        )
+
+        # embed image nodes as images directly
+        image_nodes = self._get_node_with_embedding(
+            image_nodes, show_progress, is_image=True
+        )
+        new_img_ids = self.storage_context.vector_stores[self.image_store_key].add(
+            image_nodes, **insert_kwargs
+        )
+
+        # TODO: can vector stores just store images directly? Then no need for docstore
+        # Maybe a fix for later
+
+        # if the vector store doesn't store text, we need to add the nodes to the
+        # index struct and document store
+        all_nodes = text_nodes + image_nodes
+        all_new_ids = new_text_ids + new_img_ids
+        if not self._vector_store.stores_text or self._store_nodes_override:
+            for node, new_id in zip(all_nodes, all_new_ids):
+                # NOTE: remove embedding from node to avoid duplication
+                node_without_embedding = node.copy()
+                node_without_embedding.embedding = None
+
+                index_struct.add_node(node_without_embedding, text_id=new_id)
+                self._docstore.add_documents(
+                    [node_without_embedding], allow_update=True
+                )
+        else:
+            # NOTE: if the vector store keeps text,
+            # we only need to add image and index nodes
+            for node, new_id in zip(all_nodes, all_new_ids):
+                if isinstance(node, (ImageNode, IndexNode)):
+                    # NOTE: remove embedding from node to avoid duplication
+                    node_without_embedding = node.copy()
+                    node_without_embedding.embedding = None
+
+                    index_struct.add_node(node_without_embedding, text_id=new_id)
+                    self._docstore.add_documents(
+                        [node_without_embedding], allow_update=True
+                    )
+
+    def delete_ref_doc(
+        self, ref_doc_id: str, delete_from_docstore: bool = False, **delete_kwargs: Any
+    ) -> None:
+        """Delete a document and it's nodes by using ref_doc_id."""
+        # delete from all vector stores
+        for vector_store in self._storage_context.vector_stores.values():
```

Comment:
in this case it won't delete image document?

Reply:
it will, since it will iterate over all vector stores and also call delete on the docstore and index store (although I should test this tbh)